src/maelstrom/workload/lock.clj

```diff
@@ -0,0 +1,79 @@
+(ns maelstrom.workload.lock
+  "A distributed lock workload: sends lock and unlock messages."
+  (:require [maelstrom [client :as c]]
+            [jepsen [checker :as checker]
+                    [client :as client]
+                    [generator :as gen]]
+            [schema.core :as s]))
+
+(c/defrpc lock!
+  "Clients send `lock` messages to servers. Servers should respond with
+  `lock_ok` messages when the lock is granted."
+  {:type (s/eq "lock")}
+  {:type (s/eq "lock_ok")})
+
+(c/defrpc unlock!
+  "Clients send `unlock` messages to servers. Servers should respond with
+  `unlock_ok` messages acking that the lock is released"
+  {:type (s/eq "unlock")}
+  {:type (s/eq "unlock_ok")})
+
+(defn client
+  ([net]
+   (client net nil nil))
+  ([net conn node]
+   (reify client/Client
+     (open! [this test node]
+       (client net (c/open! net) node))
+
+     (setup! [this test])
+
+     (invoke! [_ test op]
+       (c/with-errors op #{}
+         (let [rpc (case (:f op) :lock lock! :unlock unlock!)
+               timeout (* 1000 (:time-limit test))] ; timeout at most last op from each client
+           (rpc conn node {} timeout)
+           (assoc op :type :ok))))
+
+     (teardown! [_ test])
+
+     (close! [_ test]
+       (c/close! conn)))))
+
+(defn checker
+  "Expects that a lock is only given when released by previous owner."
+  []
+  (reify checker/Checker
+    (check [this test history opts]
+      (let [[_ errs]
+            (reduce
+              (fn [[acquired errors] op]
+                (case [(:type op) (:f op)]
+                  [:ok :lock]
+                  (if acquired
+                    [acquired (conj errors
+                                   (str "already acquired by process " acquired
+                                        " when lock completed at " op))]
+                    [(:process op) errors])
```

Comment:
I'm not wholly sure this is safe--it assumes the linearizable lock order is the same as the order of `ok` operations, but that might not be true. I suggest using the `linearizable` checker here instead (https://jepsen-io.github.io/jepsen/jepsen.checker.html#var-linearizable). You might need a little extra state in the model to verify that locks are released only by their acquirer, but https://github.com/jepsen-io/knossos/blob/master/src/knossos/model.clj#L92-L109 should get you started.

Reply:
Yeah, I'm really not sure this is OK. There's also no consideration of info responses, which means it's going to improperly report safety violations that don't actually exist when processes crash. I'd like to either a.) see this replaced with checker/linearizable, or b.) rewritten and accompanied by a thorough generative test suite. My guess is option a is probably more tractable.